VoodooRMI/Transports/I2C/RMII2C.cpp

```diff
@@ -0,0 +1,335 @@
+{
+    int error = 0, attempts = 0;
+
+    OSData *data;
+    data = OSDynamicCast(OSData, provider->getProperty("name"));
+    IOLog("%s: RMII2C probing %s\n", getName(), data->getBytesNoCopy());
+
+    IOService *service = super::probe(provider, score);
+    if(!service) {
+        IOLog("%s: Failed to probe provider\n", getName());
+        return NULL;
+    }
+
+    device_nub = OSDynamicCast(VoodooI2CDeviceNub, provider);
+    if (!device_nub) {
+        IOLog("%s: Could not cast nub\n", getName());
+        return NULL;
+    }
+
+    do {
+        error = rmi_set_mode(RMI_MODE_ATTN_REPORTS);
+        IOLog("%s: Trying to set mode, attempt %d\n", getName(), attempts);
+        IOSleep(500);
+    } while (error && attempts++ < 5);
+
+    if (error) {
+        IOLog("%s: Failed to set mode\n", getName());
+        return NULL;
+    }
+
+    page_mutex = IOLockAlloc();
+    IOLockLock(page_mutex);
+    /*
+     * Setting the page to zero will (a) make sure the PSR is in a
+     * known state, and (b) make sure we can talk to the device.
+     */
+    error = rmi_set_page(0);
+    IOLockUnlock(page_mutex);
+    if (error) {
+        IOLog("%s: Failed to set page select to 0\n", getName());
+        return NULL;
+    }
+    return this;
+}
+
+bool RMII2C::start(IOService *provider)
+{
+    if(!super::start(provider))
+        return false;
+
+    work_loop = getWorkLoop();
+
+    if (!work_loop) {
+        IOLog("%s: Could not get work loop\n", getName());
+        goto exit;
+    }
+
+    command_gate = IOCommandGate::commandGate(this);
+    if (!command_gate || (work_loop->addEventSource(command_gate) != kIOReturnSuccess)) {
+        IOLog("%s: Could not open command gate\n", getName());
+        goto exit;
+    }
+
+    /* Implementation of polling */
+    interrupt_source = IOInterruptEventSource::interruptEventSource(this, OSMemberFunctionCast(IOInterruptEventAction, this, &RMII2C::interruptOccured), device_nub, 0);
+    if (!interrupt_source) {
+        IOLog("%s: Could not get interrupt event source\n, trying to fallback on polling.", getName());
+        interrupt_simulator = IOTimerEventSource::timerEventSource(this, OSMemberFunctionCast(IOTimerEventSource::Action, this, &RMII2C::simulateInterrupt));
+        if (!interrupt_simulator) {
+            IOLog("%s: Could not get timer event source\n", getName());
+            goto exit;
+        }
+        work_loop->addEventSource(interrupt_simulator);
+        IOLog("%s: Polling mode initialisation succeeded.", getName());
+        setProperty("Interrupt mode", "Polling");
+    } else {
+        work_loop->addEventSource(interrupt_source);
+        IOLog("%s: running on interrupt mode.", getName());
+        setProperty("Interrupt mode", "Native");
+    }
+
+    PMinit();
+    provider->joinPMtree(this);
+    registerPowerDriver(this, RMIPowerStates, 2);
+
+    setProperty("VoodooI2CServices Supported", kOSBooleanTrue);
+    setProperty(HasResetIdentifier, kOSBooleanTrue);
```

Comment:
RMII2C does not override reset, so I would not set this property.

Reply:
It's an inline one at https://github.com/VoodooSMBus/VoodooRMI/blob/7d43f239f9c6f5c9f3e25d5a429df2b552a551e0/VoodooRMI/Transports/I2C/RMII2C.hpp#L54

Reply:
oh, missed that. The RMI4 doesn't specify a reset either so was confused, lol. This is fine